### PR TITLE
fix(test): disable preventDefault() for beforeunload event

### LINKED ISSUE #18910: [test runner]: Error messages not printed to terminal
I ran into an admittedly weird bug while working on https://github.com/denoland/fresh/pull/1139 . For some reason when _both_ an `npm:` prefixed import and a spawned child process is present, any uncaught error messages are not printed to the terminal in our test runner. The odd thing is that the npm import doesn't even have to be used and can have no side effects. Running the code without the test runner shows the error message as expected.

source:

```ts
// Commenting this unused import that has no side effects somehow
// fixes the missing error message in the test output.
// import * as kl from "npm:kolorist";

Deno.test("Foo", async () => {
  // Commenting either this or the `kolorist` import makes the
  // error message show up, but when both are present it somehow
  // doesn't print the error message
  new Deno.Command(Deno.execPath(), {
    args: ["run", "-A", "server.ts"],
    stdin: "null",
    stdout: "inherit",
    stderr: "inherit",
  }).spawn();

  // This error message isn't printed to the terminal
  throw new Error("This message should be printed to the terminal");
});
```

**Steps to reproduce:**

1. Clone https://github.com/marvinhagemeister/deno-test-uncaught-error-bug
2. Run `deno test -A foo.ts`

**Actual terminal output:**

Note how it just says "FAILED" but doesn't print the error message at all.

```sh
Check file:///Users/marvinhagemeister/dev/test/deno-bugs/test-console/foo.ts
running 1 test from ./foo.ts
Foo ... FAILED (12ms)
------- output -------
Server running: http://localhost:8300/
Listening on http://localhost:8300/
```

**Expected terminal output:**

```sh
$ deno test -A foo.ts
Check file:///Users/marvinhagemeister/dev/test/deno-bugs/test-console/foo.ts
running 1 test from ./foo.ts
Foo ... FAILED (13ms)

 ERRORS

Foo => ./foo.ts:5:6
error: Error: This message should be printed to the terminal
  throw new Error("This message should be printed to the terminal");
        ^
    at file:///Users/marvinhagemeister/dev/test/deno-bugs/test-console/foo.ts:16:9

 FAILURES

Foo => ./foo.ts:5:6

FAILED | 0 passed | 1 failed (29ms)

error: Test failed
```

### PR DESCRIPTION
**Pull request by nayeemrmn**
Fixes #18910.

### CODE REVIEW

**Comment by bartlomieju** [comment]
This seems more like a bandaid rather than fixing the underlying problem - I believe the root cause is the one described in https://github.com/denoland/deno/issues/16928 that I wanted to fix in the coming days (we now have all the ingredients to properly fix it). I'm not in favor by altering the behavior of `beforeunload` events in `deno bench`/`deno test`. What do you say I try to fix it properly with #16928 first?

**Comment by nayeemrmn** [comment]
I don't think it's a band-aid. It's possible there are _other_ issues that lead to the following block being needed: https://github.com/denoland/deno/blob/10ae5ee26557107b22524b1a84ebb56ed7d23fb4/ext/node/polyfills/process.ts#L736-L742
But while investigating it I noticed the inconsistency of whether or not the event loop is allowed to run to completion after a test run. This PR consolidates on the typical case where it doesn't, rather than only doing it if there's a `beforeunload` handler which prevents default. It happens to fix that issue.

**Comment by nayeemrmn** [comment]
The only alternative to this as I described in https://github.com/denoland/deno/issues/18910#issuecomment-1528771561 is to always call `worker.run_event_loop()` after the test loop. But I don't like that because it can lead to hanging with no info even though test results have been fully gathered.

**Comment by bartlomieju** [comment]
> I don't think it's a band-aid. It's possible there are _other_ issues that lead to the following block being needed:
> 
> https://github.com/denoland/deno/blob/10ae5ee26557107b22524b1a84ebb56ed7d23fb4/ext/node/polyfills/process.ts#L736-L742
> 
> 
> But while investigating it I noticed the inconsistency of whether or not the event loop is allowed to run to completion after a test run. This PR consolidates on the typical case where it doesn't, rather than only doing it if there's a `beforeunload` handler which prevents default. It happens to fix that issue.

Not really - there are no other issues that led to this block looking like it does, besides I had no other way to polyfill this event when this code lived in `std/node`. I'm fairly certain that I can rework it to not clash with "beforeunload".

> At the end of the test run, it's true that there is more work in the event loop because the child process is still open. But for normal test runs we don't attempt to run the event loop unless a JS beforeunload handler calls e.preventDefault() (which the Node compat layer does I assume for good reason). This is kind of strange.

Again, there was no good reason, besides the fact that I wasn't able to polyfill it in any other way at the time of writing that code, which as the linked issue described is actually buggy.

> We should either attempt to complete the event loop before dispatching beforeunload no matter what, or be ruthless and not allow the user to keep a worker alive once the tests from that module are finished i.e. ignore the defaultPrevented for beforeunload in test runs. I'm strongly leaning to the latter.

Maybe, still it feels a little strange that in `deno run` one can keep the event loop alive with `beforeunload` handler as long as needed/wanted but `deno test` and `deno bench` would prevent that. 

@dsherret @lucacasonato thoughts on this?

**Comment by nayeemrmn** [comment]
> there are no other issues that led to this block looking like it does

I was referring to #18910. Does the reproduction you're using for #16928 also use `deno test`? Otherwise I do think it's unrelated to this change and can continue to be tracked there.

This is related to a separate issue of whether or not to allow the event loop to complete in a test module after finishing tests. Currently `deno test` is not consistent with `deno run`. To make it clearer, the worker clean up code should either look like this:
```rs
  // Run event loop to completion, same as `deno run`.
  worker.run_event_loop(false).await?;
  loop {
    if !worker.dispatch_beforeunload_event(located_script_name!())? {
      break;
    }
    worker.run_event_loop(false).await?;
  }
  worker.dispatch_unload_event(located_script_name!())?;
```
Or like this:
```rs
  // Don't run event loop, run unload events but don't allow them to prevent default.
  !worker.dispatch_beforeunload_event(located_script_name!())?;
  worker.dispatch_unload_event(located_script_name!())?;
```
But currently it looks like this:
```rs
  // Don't run event loop typically, run unload events and run the event loop if they prevent default.
  loop {
    if !worker.dispatch_beforeunload_event(located_script_name!())? {
      break;
    }
    worker.run_event_loop(false).await?;
  }
  worker.dispatch_unload_event(located_script_name!())?;
```
I think the first one is disruptive because existing code which might rely on the worker and its open resources being dropped after tests are finished would start to hang with no info. The second is what would normally happen currently but with the edge case smoothed off. Thoughts?

**Comment by nayeemrmn** [comment]
> Wouldn't the first case immediately break the loop once `beforeunload` event is disptached and then dispatch `unload` event?. To be honest I don't fully understand why there's a difference how these loops look between `deno run`/`deno bench`/`deno test`.

The relevant line is the invariant `worker.run_event_loop()` added at the top. Currently that's not there.

**Comment by bartlomieju** [comment]
> > Wouldn't the first case immediately break the loop once `beforeunload` event is disptached and then dispatch `unload` event?. To be honest I don't fully understand why there's a difference how these loops look between `deno run`/`deno bench`/`deno test`.
> 
> The relevant line is the invariant `worker.run_event_loop()` added at the top. Currently that's not there.

Exactly :D how does it even work currently if we're not polling the event loop? Am I missing something simple here? I guess it's because of `worker.js_runtime.call_and_await`? But in such case if there's a setup code at top level that uses async APIs it's not polled before the first test case runs, right?

**Comment by nayeemrmn** [comment]
> I guess it's because of `worker.js_runtime.call_and_await`? But in such case if there's a setup code at top level that uses async APIs it's not polled before the first test case runs, right?

Yep, the same was true since `runTests()` in JS. There's no opportunity to run the event loop until awaiting the first test result and you can't do `worker.run_event_loop()` before running tests because it would never resolve if say you started a server for testing.

**Comment by bartlomieju** [comment]
> > I guess it's because of `worker.js_runtime.call_and_await`? But in such case if there's a setup code at top level that uses async APIs it's not polled before the first test case runs, right?
> 
> Yep, the same was true since `runTests()` in JS. There's no opportunity to run the event loop until awaiting the first test result and you can't do `worker.run_event_loop()` before running tests because it would never resolve if say you started a server for testing.

Okay, that makes sense. Sorry for back and forth on this.

File: cli/tools/bench.rs [hunk 1]
```diff
@@ -497,12 +497,7 @@ async fn bench_specifier(
     sender.send(BenchEvent::Result(desc.id, result))?;
   }
 
-  loop {
-    if !worker.dispatch_beforeunload_event(located_script_name!())? {
-      break;
-    }
-    worker.run_event_loop(false).await?;
-  }
+  worker.dispatch_beforeunload_event(located_script_name!())?;
```
Comment (bartlomieju) [inline cli/tools/bench.rs:500]: Please add a comment here and in `test.rs` why we're disregarding the `e.preventDefault()` here

**Review by bartlomieju — COMMENTED** [review]
LGTM, once the comments are added

**Review by bartlomieju — APPROVED** [review]
LGTM

**Pull request merged**